example/src/main/java/com/thefuntasty/mvvmsample/ui/coroutinesresult/CoroutinesResultViewModel.kt

```diff
@@ -0,0 +1,73 @@
+package com.thefuntasty.mvvmsample.ui.coroutinesresult
+
+import com.thefuntasty.mvvm.crusecases.BaseCrViewModel
+import com.thefuntasty.mvvm.crusecases.getOrCancel
+import com.thefuntasty.mvvm.crusecases.getOrElse
+import com.thefuntasty.mvvm.crusecases.getOrThrow
+import com.thefuntasty.mvvm.crusecases.map
+import com.thefuntasty.mvvm.crusecases.recover
+import com.thefuntasty.mvvmsample.domain.dummy.ConfirmDataSavedSuccessfullyUseCase
+import com.thefuntasty.mvvmsample.domain.dummy.GetDataFromDeviceUseCase
+import com.thefuntasty.mvvmsample.domain.dummy.SaveDataToFirstServerUseCase
+import com.thefuntasty.mvvmsample.domain.dummy.SaveDataToSecondServerUseCase
+import kotlinx.coroutines.delay
+import javax.inject.Inject
+
+class CoroutinesResultViewModel @Inject constructor(
+    private val getDataFromDeviceUseCase: GetDataFromDeviceUseCase,
+    private val saveDataToFirstServerUseCase: SaveDataToFirstServerUseCase,
+    private val saveDataToSecondServerUseCase: SaveDataToSecondServerUseCase,
+    private val confirmDataSavedSuccessfullyUseCase: ConfirmDataSavedSuccessfullyUseCase,
+    override val viewState: CoroutinesResultViewState
+) : BaseCrViewModel<CoroutinesResultViewState>() {
+
+    fun onStartLoadingClicked() = launchWithHandler {
+        showLoading()
+
+        // If the use case fails then result of `getOrElse` is returned
+        val deviceData: String = getDataFromDeviceUseCase.execute().getOrElse { "Default data" }
+        setLoadingState(step = "1")
+
+        // If the use case fails then `showError` is called and the coroutine is canceled with CancellationException
+        val firstSave: String = saveDataToFirstServerUseCase.execute(deviceData).getOrCancel { showError(step = "2") }
+        setLoadingState(step = "2")
+
+        // If the use case fails then result of `recover` is returned
+        val secondSave = saveDataToSecondServerUseCase.execute(deviceData).map { "OK" }.recover { "Ignored error" }.getOrThrow()
+        setLoadingState(step = "3")
+
+        // The use case returns either (result, null) or (null, Throwable)
+        val (result, _) = confirmDataSavedSuccessfullyUseCase.execute(firstSave to secondSave)
+
+        setLoadingState(step = "4")
+        delay(500)
```

Comment:
<table data-meta="generated_by_danger-pr">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/thefuntasty/mvvm-android/blob/1e22aaaf9def5bbeda391f2cd9735be2fc018fbd/example/src/main/java/com/thefuntasty/mvvmsample/ui/coroutinesresult/CoroutinesResultViewModel.kt#L43"></span>This expression contains a magic number. Consider defining it to a well named constant.</td>
    </tr>
  </tbody>
</table>